File: spice2/Equation.cpp
//
// Created by rhys on 18/04/2021.
//

#include <iostream>
#include "Equation.h"

Term::Term(double coefficient, Unknown *variable) {
    this->coefficient = coefficient;
    this->variable = variable;
}

std::string Term::str() {
    std::string prefix = coefficient == 1? "" : coefficient == -1? "-" : std::to_string(coefficient);

    return prefix + variable->toTermName();
}

UnknownCurrent::UnknownCurrent(MNAElement *element) : Unknown(CURRENT) {
    this->element = element;
}

std::string UnknownCurrent::toTermName() {
    return "I"+std::to_string(element->n0)+"_"+std::to_string(element->n1);
}

bool UnknownCurrent::equals(Unknown* other) {
    if(other->t != Unknown::CURRENT){
        return false;
    }

    auto* c = (UnknownCurrent*) other;
    return c->element == element;
}

UnknownVoltage::UnknownVoltage(int node)  : Unknown(VOLTAGE) {
    this->node = node;
}

std::string UnknownVoltage::toTermName() {
    return "V"+std::to_string(node);
}

bool UnknownVoltage::equals(Unknown *other) {
    if(other->t != Unknown::VOLTAGE){
        return false;
    }

    auto* c = (UnknownVoltage*) other;
    return c->node == node;
}

Equation::Equation(double value, std::vector<Term*> terms) {
    this->value = value;
    this->terms = terms;
}

void Equation::stamp(int row, Eigen::MatrixXd* a, Eigen::MatrixXd* z, std::function<int(Unknown*)> getIndexFunc) {
    (*z)(row, 0) = value;

    for(auto t : terms){
        int index = getIndexFunc(t->variable);
        (*a)(row, index) = t->coefficient + (*a)(row, index);
    }
}

std::string Equation::str() {
    std::stringstream termList;

    for(auto t : terms){
        termList << t->str() << " +";
    }

    return "{"+termList.str()+"} = " + std::to_string(value);
}

Unknown::Unknown(Type t) {
    this->t = t;
}

File: spice3/Equation.py
import numpy as np

from Unknowns import UnknownVoltage


class Term:
    def __init__(self, coefficient: float, variable):
        self.coefficient = coefficient
        self.variable = variable

    def __str__(self):
        prefix = "-" if self.coefficient == 0 else ""

        return prefix + str(self.variable)

class Equation:
    def __init__(self, value: float, terms: list[Term]):
        self.value = value
        self.terms = terms

    def stamp(self, row: int, A: np.array, z: np.array, get_index: callable):
        z[row, 0] = self.value
        print("\t Stamp {} onto z at [0, {}]\n".format(self.value, row))

        for t in self.terms:
            index = get_index(t.variable)
            rhs = t.coefficient + A[row, index]
            print("\t\tStamp {} {} onto A:".format(type(t.variable), "Node "+str(t.variable.node) if type(t.variable) == UnknownVoltage else t.variable.element))
            print("\t\t\tCoordinates = [{}, {}]\n\t\t\tExisting = {}\n\t\t\tAdding = {}\n\t\t\tNew = {}".format(index, row, A[row, index], t.coefficient, rhs), end="\n\n\n")
            A[row, index] = rhs

    def __str__(self):
        term_list = []

        for t in self.terms:
            term_list.append(str(t))

        result = f"{'+'.join(term_list)}={self.value}"

        return result.replace("\\+\\-", "\\-")